File: LeetCode/1005/main.py
"""
DIFFICULTY : easy
TAGS : array, greedy, sorting
"""


class Solution:
    def largestSumAfterKNegations(self, arr: List[int], k: int) -> int:
        arr.sort()
        i = 0
        while (i < len(arr) and i < k and arr[i] < 0):
            arr[i] = -arr[i]
            i += 1
        return sum(arr) - (k - i) % 2 * min(arr) * 2

File: LeetCode/1005/main.cpp
class Solution {
public:
  int largestSumAfterKNegations(vector<int> &arr, int k) {

    int n = arr.size();

    sort(arr.begin(), arr.end());

    for (int i = 0; i < n; i++) {
      if (arr[i] < 0 && k > 0) {
        arr[i] = -arr[i];
        k--;
      }
    }

    sort(arr.begin(), arr.end());

    if (k % 2 == 1) {
      arr[0] = -arr[0];
    }

    return accumulate(arr.begin(), arr.end(), 0);
  }
};